File: deferDemo/deferDemo/ViewController.m
//
//  ViewController.m
//  deferDemo
//
//  Created by kfw on 2019/6/19.
//  Copyright © 2019 kfw. All rights reserved.
//

#import "ViewController.h"


/*
 函数执行完最后执行。 与swift defer 函数一样。
 示例：
 OnBlockExit(block_1) {
    NSLog(@"OnBlockExit 1");
 };
 */
//#ifdef __GNUC__
//__unused static void cleanUpBlock(__strong void(^*block)(void)) {
//    (*block)();
//}
//#define OnBlockExit(block_name) __strong void(^block_name)(void) __attribute__((cleanup(cleanUpBlock), unused)) = ^
//#endif


void cleanup_block(int *a) {
    printf("%d\n", *a);
}



@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}


- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    
    int variable __attribute__((cleanup(cleanup_block))) = 2;
    
//    OnBlockExit(block_1) {
//        NSLog(@"OnBlockExit 1");
//    };
//    OnBlockExit(block_2) {
//        NSLog(@"OnBlockExit 2");
//    };
//    OnBlockExit(block_3) {
//        NSLog(@"OnBlockExit 3");
//    };
//    OnBlockExit(w) {
//        NSLog(@"OnBlockExit 4");
//    };
//
//    NSLog(@"End");
}

@end

File: SwiftDemo/SwiftDemo11-析构过程/SwiftDemo11-析构过程/ViewController.swift
//
//  ViewController.swift
//  SwiftDemo11-析构过程
//
//  Created by kfw on 2019/7/12.
//  Copyright © 2019 kfw. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
// 析构器是在实例释放发生前被自动调用的。你不能主动调用析构器。子类继承了父类的析构器，并且在子类析构器实现的最后，父类的析构器会被自动调用。即使子类没有提供自己的析构器，父类的析构器也同样会被调用。
    deinit {
        // 执行析构过程
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }


}

